pkg/vgmanager/lvm.go

```diff
@@ -23,33 +23,206 @@ import (
 )
 
+type lvmError string
+
+func (s lvmError) Error() string { return string(s) }
+
+var (
+	ErrVolumeGroupNotFound = lvmError("volume group not found")
+)
+
 const (
-	lvmPath = "/usr/sbin/lvm"
+	lvmCmd      = "/usr/sbin/lvm"
+	vgCreateCmd = "/usr/sbin/vgcreate"
+	vgExtendCmd = "/usr/sbin/vgextend"
+	vgRemoveCmd = "/usr/sbin/vgremove"
+	pvRemoveCmd = "/usr/sbin/pvremove"
 )
 
+// vgsOutput represents the output of the `vgs --reportformat json` command
+type vgsOutput struct {
+	Report []struct {
+		Vg []struct {
+			Name string `json:"vg_name"`
+		} `json:"vg"`
+	} `json:"report"`
+}
+
+// pvsOutput represents the output of the `pvs --reportformat json` command
+type pvsOutput struct {
+	Report []struct {
+		Pv []struct {
+			Name   string `json:"pv_name"`
+			VgName string `json:"vg_name"`
+		} `json:"pv"`
+	} `json:"report"`
+}
+
 // VolumeGroup represents a volume group of linux lvm.
 type VolumeGroup struct {
+	// Name is the name of the volume group
 	Name string `json:"vg_name"`
+
+	// PVs is the list of physical volumes associated with the volume group
+	PVs []string `json:"pvs"`
 }
 
-// ListVolumeGroups lists all volume groups.
-func ListVolumeGroups(exec internal.Executor) ([]VolumeGroup, error) {
-	// volumeGroupMap type is used for marshalling the `vgs --reportformt json` command output
-	// Sample output:
-	// `{"report": [ {"vg": [{"vg_name":"fedora", "pv_count":"1", "lv_count":"3", "snap_count":"0", "vg_attr":"wz--n-", "vg_size":"<475.94g", "vg_free":"0 "}]}]`
-	volumeGroupMap := map[string][]map[string][]VolumeGroup{}
-	arg := []string{
+// Create creates a new volume group
+func (vg VolumeGroup) Create(exec internal.Executor, pvs []string) error {
+	if vg.Name == "" {
+		return fmt.Errorf("failed to create volume group. Volume group name is empty")
+	}
+
+	if len(pvs) == 0 {
+		return fmt.Errorf("failed to create volume group. Physical volume list is empty")
+	}
+
+	args := []string{vg.Name}
+	args = append(args, pvs...)
+
+	_, err := exec.ExecuteCommandWithOutputAsHost(vgCreateCmd, args...)
+	if err != nil {
+		return fmt.Errorf("failed to create volume group %q. %v", vg.Name, err)
+	}
+
+	return nil
+}
+
+// Extend extends the volume group only if new physical volumes are available
+func (vg VolumeGroup) Extend(exec internal.Executor, pvs []string) error {
+	if vg.Name == "" {
+		return fmt.Errorf("failed to extend volume group. Volume group name is empty")
+	}
+
+	if len(pvs) == 0 {
+		return fmt.Errorf("failed to extend volume group. Physical volume list is empty")
+	}
+
+	args := []string{vg.Name}
+	args = append(args, pvs...)
+
+	_, err := exec.ExecuteCommandWithOutputAsHost(vgExtendCmd, args...)
+	if err != nil {
+		return fmt.Errorf("failed to extend volume group %q. %v", vg.Name, err)
+	}
+
+	return nil
+}
+
+// Delete deletes a volume group and the physical volumes associated with it
+func (vg VolumeGroup) Delete(exec internal.Executor) error {
+	// Remove Volume Group
+	vgArgs := []string{vg.Name}
```

Comment:
- use `--nolocking` as well? rationale is if the volume group is huge acquiring lock takes time